### ISSUE TITLE
[android] Zombie .Torrents

### ISSUE THREAD

**Issue by gubatron** [frostwire/frostwire#620]
Torrents no longer there keep appearing, as if the results from the Content provider were cached.
We should refresh such output if a file has been detected as no longer there, very annoying behavior trying to open .torrents that are no longer in disk.

**Comment by gubatron** [comment]
@aldenml since you were taking a clean at Table Fetchers, please take a look at this ticket

**Comment by gubatron** [comment]
ping @aldenml

**Comment by aldenml** [comment]
ACK

**Comment by aldenml** [comment]
@muckachina are you able to reproduce this issue in any of your devices?

**Comment by muckachina** [comment]
If is this issue what you are mentioning here:

https://github.com/frostwire/qa-android/issues/102

I've been testing it today and I can't reproduce with some fresh downloads but still happens with old downloads that I try to 'remove torrent and data' from Transfer, they keep appearing on My Files > My Torrents

**Comment by gubatron** [comment]
ping @votaguz (where you able to take a look at this?)

**Comment by gubatron** [comment]
This ticket is related to this
https://github.com/frostwire/frostwire/issues/274

**Comment by gubatron** [comment]
wonder if the `getContentUri()` method on the torrent table fetcher should use some other logic.
I can't replicate when running from source any more (without SD Card)

will update if It works fine running from a signed .apk

**Comment by aldenml** [comment]
I was finally able to replicate this issue, and should be fixed with https://github.com/frostwire/frostwire/commit/624ccaf027e8e4715f73b4d2e94d1eab2e849120

@muckachina give it a try and let me know

**Comment by gubatron** [comment]
![jean-claude-dancing](https://user-images.githubusercontent.com/163977/36697108-56efd126-1b03-11e8-9ae7-305d46a7c185.gif)

**Comment by muckachina** [comment]
_Tested on FW 2.0.2-build-497 on android 6 and 7
Files saved within SD card and Internal Memory_

I cannot reproduce it again with the files I have in Transfer... The 'zombie' torrents that remained there before were from old builds, but since I don't have old data in Transfer for the latest builds I could say is fixed.

@gubatron Let me know If you still find the issue with old downloads or any resent downloads, I closed the issue here https://github.com/frostwire/frostwire/issues/274#event-907814800 unless I can find a particular scenario where I can reproduce.

**Comment by gubatron** [comment]
I still find the issue :(

Just wait a few minutes and open the app again and all the torrents are there.

**Comment by muckachina** [comment]
Ok @gubatron but what you see again are .torrent files you selected from 'My Torrents' and 'Delete' them right? that's another issue, I closed 'Delete Torrent and Data' from Transfer list

Actually 'Delete' .torrents has this ticket open: https://github.com/frostwire/qa-android/issues/89 but the scenario is changing again, I'll updated in few minutes after checking all cases, now when 'Deleted' a file that was at the internal Memory the file disappears from My Torrents list but remains saved in the Internal Memory, when I Exit and Open FrostWire again as you say the torrent appears on My Torrents list again.

**Comment by muckachina** [comment]
So, basically this is the current status of 'Delete' at 'My Files':

### **A) Deleting a file from 'My Files' (Files saved on the Internal Memory)**

**BUG** --> .torrent files are not being deleted from _Internal Memory > FrostWire > 'Torrents'_ folder

**To reproduce:**
1) Go to My Files > My Torrents
2) Select a .torrent file and click the icon 'Delete' > The file will disappear from My Torrents list
3) Exit FW and check with a File Explorer at the _Internal Memory > FrostWire > Torrents_ folder that the .torrent file is still there
4) Open FW again and you will see the file re-appears at _My Files > My Torrents_ list

The rest of files saved at the _Internal Memory > 'Torrents Data'_ are 'Deleted' properly from My Files and _'Torrents Data'_ folder

### **B) Deleting a file from 'My Files' (Files saved on the SD card and Storage Location pointing there)**

**BUG** --> .torrent files are not being deleted from _Internal Memory > FrostWire > 'Torrents'_ folder even when the Storage Location is pointing to that folder on the SD card

**To reproduce:**
1) Go to Storage Location and make sure you are pointing the SD card folder where it's placed the file you want to delete
2) Go to My Files > My Torrents and select a .torrent file then click the icon 'Delete' > The file will disappear from My Torrents list
3) Exit FW and check with a File Explorer at the _SD Card > Folder-Selected > Torrents_ folder that the .torrent file is still there
4) Open FW again and you will see the file re-appears at _My Files > My Torrents_ list

Delete files saved within 'Torrents Data' folder works properly **only when the Storage Location is pointing the SD card specific folder where the files are saved**

### **C) Deleting a file from 'My Files' (Files saved on the SD card and Storage Location pointing to a different folder no where the file is saved)**

**BUG** --> The 'Delete' option does not work for any file saved within the SD card when the Storage Location is pointing to a different folder

**To reproduce:**
1) Go to Storage Location and make sure you are pointing a different folder where it's placed the file you want to delete
2) Go to My Files and select any file then click the icon 'Delete' > The file will disappear from My Files list
3) Exit FW and check with a File Explorer at the SD Card that the file you deleted is still there
4) Open FW again and you will see the file re-appears at My Files (It could take a time till FW re-scans the phone again)

Updated here https://github.com/frostwire/qa-android/issues/89

**Comment by gubatron** [comment]
@aldenml 
I believe I just found the reason why this isn't working.

![screen shot 2018-03-02 at 2 10 47 pm](https://user-images.githubusercontent.com/163977/36922018-ad6a4842-1e23-11e8-99f3-ea4bf6bba8b1.png)

`BTEngine.readTorrentPath()` is not returning a valid .torrent file.

It does find the torrent inside the internal libtorrent folder, but it's not finding the .torrent inside the user facing folder. When I ask the returning file if it exists, it returns `false`.

Maybe there's a bug here when the resume data key with the original path is being written, I wouldn't really know now

![screen shot 2018-03-02 at 2 14 05 pm](https://user-images.githubusercontent.com/163977/36922121-188097f8-1e24-11e8-8a19-1265851a458e.png)

This is what it's retrieving as a File object, let's see the path:
![screen shot 2018-03-02 at 2 15 29 pm](https://user-images.githubusercontent.com/163977/36922157-2eb765d8-1e24-11e8-8a1b-fa5f3e80a447.png)

`/storage/emulated/0/FrostWire/Torrents/FROSTCLICK_FROSTWIRE_CREATIVE_COMMONS_MIXTAPE_VOL_6__MP3__December_2017.torrent`

Taking a closer look

**Comment by gubatron** [comment]
nvm, I had checked "Delete .torrents when downloads are started"

**Comment by gubatron** [comment]
@muckachina this should be fixed for build `499`

**Comment by muckachina** [comment]
@gubatron Since this message you launched on the web 5 new builds (one per day) so, finally testing on build 504 and updating issue:

A) **FIXED**
.torrent files are not being deleted from Internal Memory > FrostWire > 'Torrents' folder

B) **FIXED**
.torrent files are not being deleted from Internal Memory > FrostWire > 'Torrents' folder even when the Storage Location is pointing to that folder on the SD card

C) **PERSISTS**
**BUG** --> The 'Delete' option does not work for any file saved within the SD card when the Storage Location is pointing to a different folder

To reproduce:

1) Go to Storage Location and make sure you are pointing a different folder where it's placed the file you want to delete
2) Go to My Files and select any file then click the icon 'Delete' > The file will disappear from My Files list
Exit FW and check with a File Explorer at the SD Card that the file you deleted is still there
3) Open FW again and you will see the file re-appears at My Files (It could take a time till FW re-scans the phone again)

So basically we are back to the initial issue where files on SD card not always can be deleted. This apply to .torrent and other files.

**Comment by gubatron** [comment]
Yeah, let's keep that a separate issue: `Can't delete files on SD Card`.
